In [389]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [390]:
df=pd.read_csv("C:\jupytercsv\student.csv")
df.head()

Roll     Adm  Std  Sec Age Gen Place  Faage  Faedu  Faocc  ...  Work  \
0  6A001  1416.0    6    1  11   2    23   51.0      5      3  ...    No   
1  6A002  1425.0    6    1  11   2     1   38.0      6      2  ...    No   
2  6A003  1417.0    6    1  10   2    34   44.0      5      2  ...    No   
3  6A004  1410.0    6    1  11   2     5   37.0      6      3  ...    No   
4  6A005  1421.0    6    1  11   2    23   36.0      7      6  ...    No   

   Travel  Tran Phy  Men  Sleep Goout  Phone  Somed  Phohr  
0      55     2   4    5    7.0     0      2      1      0  
1      15     4   5    5    9.0     1      2      1      0  
2       1     2   5    5    9.0     1      2      1      0  
3       5     2   3    2    8.0     2      2      2      1  
4      30     2   5    5    8.0     1      1      3      1  

[5 rows x 47 columns]

In [391]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 47 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Roll      350 non-null    object 
 1   Adm       323 non-null    float64
 2   Std       350 non-null    int64  
 3   Sec       350 non-null    int64  
 4   Age       350 non-null    object 
 5   Gen       350 non-null    object 
 6   Place     350 non-null    object 
 7   Faage     349 non-null    float64
 8   Faedu     350 non-null    int64  
 9   Faocc     350 non-null    int64  
 10  Fasal     350 non-null    int64  
 11  Moage     350 non-null    int64  
 12  Moedu     350 non-null    int64  
 13  Moocc     350 non-null    object 
 14  Mosal     350 non-null    int64  
 15  Stay      350 non-null    int64  
 16  Famtyp    347 non-null    object 
 17  Count     350 non-null    int64  
 18  Pamas     350 non-null    int64  
 19  Famrel    350 non-null    int64  
 20  Loved     350 non-null    int64 

In [392]:
df.describe(include='all')

Roll          Adm         Std         Sec  Age  Gen Place  \
count     350   323.000000  350.000000  350.000000  350  350   350   
unique    343          NaN         NaN         NaN   12    5    31   
top     9C018          NaN         NaN         NaN   16    1     1   
freq        2          NaN         NaN         NaN   60  214   113   
mean      NaN  1109.817337    8.882857    1.897143  NaN  NaN   NaN   
std       NaN   347.672109    2.046870    0.783602  NaN  NaN   NaN   
min       NaN   523.000000    6.000000    1.000000  NaN  NaN   NaN   
25%       NaN   682.500000    7.000000    1.000000  NaN  NaN   NaN   
50%       NaN  1229.000000    9.000000    2.000000  NaN  NaN   NaN   
75%       NaN  1403.000000   11.000000    3.000000  NaN  NaN   NaN   
max       NaN  1572.000000   12.000000    3.000000  NaN  NaN   NaN   

             Faage       Faedu       Faocc  ...  Work  Travel        Tran  \
count   349.000000  350.000000  350.000000  ...   350     342  350.000000   
unique         NaN         NaN         NaN  ...     6      21         NaN   
top            NaN         NaN         NaN  ...    No      15         NaN   
freq           NaN         NaN         NaN  ...   307      77         NaN   
mean     44.217765    5.380000    3.428571  ...   NaN     NaN    2.177143   
std       6.635729    1.726633    1.651356  ...   NaN     NaN    0.988497   
min      -1.000000   -1.000000   -1.000000  ...   NaN     NaN    1.000000   
25%      41.000000    5.000000    2.000000  ...   NaN     NaN    2.000000   
50%      44.000000    5.000000    3.000000  ...   NaN     NaN    2.000000   
75%      48.000000    6.000000    5.000000  ...   NaN     NaN    2.000000   
max      61.000000   10.000000    7.000000  ...   NaN     NaN    5.000000   

               Phy         Men       Sleep       Goout       Phone      Somed  \
count   350.000000  350.000000  349.000000  350.000000  350.000000  350.00000   
unique         NaN         NaN         NaN         NaN         NaN        NaN   
top            NaN         NaN         NaN         NaN         NaN        NaN   
freq           NaN         NaN         NaN         NaN         NaN        NaN   
mean      4.200000    4.105714    7.979943    1.211429    1.714286    1.80000   
std       1.189736    1.267953    0.998360    1.921343    0.452401    1.21594   
min       1.000000    1.000000    6.000000    0.000000    1.000000    1.00000   
25%       4.000000    3.000000    7.000000    0.000000    1.000000    1.00000   
50%       5.000000    5.000000    8.000000    0.000000    2.000000    1.00000   
75%       5.000000    5.000000    8.000000    1.000000    2.000000    3.00000   
max       5.000000    5.000000   12.000000    7.000000    2.000000    5.00000   

        Phohr  
count     350  
unique     13  
top         0  
freq      152  
mean      NaN  
std       NaN  
min       NaN  
25%       NaN  
50%       NaN  
75%       NaN  
max       NaN  

[11 rows x 47 columns]

In [393]:
df.shape

(350, 47)

## Let's do some cleaning, shall we? 

##### There are 350 rows and 47 columns. 
First we'll check if a column has to be cleaned and then we'll clean it.
Starting with roll number.

In [394]:
df.Roll.describe()

count       350
unique      343
top       9C018
freq          2
Name: Roll, dtype: object

We can see that there are duplicates in roll number. As it is the primary key here we have to remove the duplicates in it.
Some students might have filled the form twice. So, we'll take the recent one into account and drop the other ones.

In [395]:
df[df["Roll"].duplicated(keep="last")]

Roll     Adm  Std  Sec Age Gen Place  Faage  Faedu  Faocc  ...  Work  \
73     7B008  1572.0    7    2  13   1    29   45.0      6      2  ...   Yes   
144    8B021  1267.0    8    2  13   1    32   37.0      5      3  ...    No   
175    9B002  1208.0    9    2  15   1     1   41.0      7      3  ...    No   
198    9C018  1291.0    9    3  14   1     1   38.0      7      2  ...    No   
282  11A2011   657.0   11    2  16   2     1   46.0      5      5  ...    No   
296   11C007   685.0   11    3  16   1    20   42.0      5      3  ...    No   
341  12A2030   585.0   12    2  16   1     1   41.0      3      1  ...   Yes   

     Travel  Tran Phy  Men  Sleep Goout  Phone  Somed  Phohr  
73       30     2   5    5    8.0     1      2      1      0  
144      45     2   5    5    8.0     0      2      1      0  
175       8     4   5    5    9.0     1      2      1      0  
198      10     1   4    4   10.0     1      2      4      3  
282      15     1   5    5    8.0     0      2      1      0  
296      15     2   5    5    9.0     0      2      3      2  
341      10     1   5    5    7.0     0      1      3      1  

[7 rows x 47 columns]

There are 7 duplicate rows.  

In [396]:
df=df.drop_duplicates(subset="Roll",keep="last").reset_index(drop=True)
df.Roll.describe()

count       343
unique      343
top       6A001
freq          1
Name: Roll, dtype: object

We have successfully dropped all the duplicates from the dataframe. Now the total number of rows are 343 and not 350.

Now we can just drop admission number column as we have roll number as student identifier and only 316 values are there in the column.

In [397]:
df.Adm.describe()

count     316.000000
mean     1111.411392
std       347.231694
min       523.000000
25%       683.250000
50%      1230.500000
75%      1405.500000
max      1572.000000
Name: Adm, dtype: float64

But before that we have to create a column called 'Fee' which is basically a binary column used to tell whether a student paid his/her previous year fee. For this we have to use another csv file 'last_year_fee_paid.csv' which contains the admission numbers of all the students who didn't pay the previous year fee

In [398]:
df2=pd.read_csv('C:\jupytercsv\last_year_fee_paid.csv')
df2.head()

Admn No.
0      1322
1      1318
2      1304
3      1400
4      1401

Now we'll create a new column 'Fee'

In [399]:
df2.dtypes

Admn No.    int64
dtype: object

In [400]:
df['Fee']=df.Adm.apply(lambda x: '0' if x in df2['Admn No.'].values else '1')
df

Roll     Adm  Std  Sec Age                       Gen Place  Faage  \
0      6A001  1416.0    6    1  11                         2    23   51.0   
1      6A002  1425.0    6    1  11                         2     1   38.0   
2      6A003  1417.0    6    1  10                         2    34   44.0   
3      6A004  1410.0    6    1  11                         2     5   37.0   
4      6A005  1421.0    6    1  11                         2    23   36.0   
..       ...     ...  ...  ...  ..                       ...   ...    ...   
338  12A2034   589.0   12    2  17                         1     5   49.0   
339  12A2035   590.0   12    2  16                         1    10   47.0   
340  12A2036   591.0   12    2  16                         1    22   49.0   
341  12A2066   576.0   12    2  17  endiraniselvam@gmail.com    28   43.0   
342   12C007   598.0   12    3  17                         1    21   45.0   

     Faedu  Faocc  ...  Travel  Tran  Phy Men  Sleep  Goout Phone  Somed  \
0        5      3  ...      55     2    4   5    7.0      0     2      1   
1        6      2  ...      15     4    5   5    9.0      1     2      1   
2        5      2  ...       1     2    5   5    9.0      1     2      1   
3        6      3  ...       5     2    3   2    8.0      2     2      2   
4        7      6  ...      30     2    5   5    8.0      1     1      3   
..     ...    ...  ...     ...   ...  ...  ..    ...    ...   ...    ...   
338      4      2  ...      10     1    5   5    8.0      0     1      2   
339      7      2  ...      35     2    4   4    7.0      1     1      4   
340      5      2  ...      25     2    4   3    8.0      0     1      4   
341      4      5  ...      15     2    5   1    8.0      0     2      1   
342      5      3  ...      15     2    5   5    9.0      1     1      1   

     Phohr  Fee  
0        0    1  
1        0    1  
2        0    1  
3        1    1  
4        1    1  
..     ...  ...  
338      1    1  
339      2    1  
340      2    0  
341      1    0  
342      1    0  

[343 rows x 48 columns]

In [401]:
df.Fee.unique()

array(['1', '0'], dtype=object)

A new column 'Fee' has been created and corresponding values have been entered. if the student has paid the fee then the 'Fee' column contains 1 else it contains 0.

Since this fee data has been gathered from previous year, The present 6th grade students will all have 1 in the 'Fee' column.

Now we can drop the 'Adm' column from the 'df' dataframe since it's of no use.

In [402]:
df=df.drop(columns="Adm")
df.shape

(343, 47)

Standard and Section:

In [403]:
df.Std.describe()

count    343.000000
mean       8.868805
std        2.051313
min        6.000000
25%        7.000000
50%        9.000000
75%       11.000000
max       12.000000
Name: Std, dtype: float64

In [404]:
df.Sec.describe()

count    343.000000
mean       1.889213
std        0.786937
min        1.000000
25%        1.000000
50%        2.000000
75%        3.000000
max        3.000000
Name: Sec, dtype: float64

Age

Wondered why Age column was showing object datatype and got the unique values.

In [405]:
df.Age.unique()

array(['11', '10', '12', '13', '21.03 2012,    12', '14', '15', '16',
       '14 yrs 11 months', '17', '16 year', '18'], dtype=object)

This column has values such as '14 yrs 11 months','21.03 2012,    12' & '16 year' which can we grouped with other values i.e. '12,'14' & '16' 

In [406]:
dic= {
    '21.03 2012,    12':'12',
    '14 yrs 11 months':'14',
    '16 year':'16'
}
df.replace(dic,inplace=True)
print(df.Age.unique())

df.Age.describe()

['11' '10' '12' '13' '14' '15' '16' '17' '18']


count     343
unique      9
top        16
freq       58
Name: Age, dtype: object

Unnecessary values have been replaced successfully. Now we can do the same thing with Gender column.

In [407]:
df.Gen.unique()

array(['2', '1', 'Male', 'Female', 'endiraniselvam@gmail.com'],
      dtype=object)

In [408]:
dic={
    'Male':'1', 'Female':'2', 'endiraniselvam@gmail.com':'2'
}
df.replace(dic,inplace=True)
print(df.Gen.unique())


['2' '1']


In [409]:
df.Gen.describe()

count     343
unique      2
top         1
freq      210
Name: Gen, dtype: object

Place

In [410]:
df.Place.describe()

count     343
unique     31
top         1
freq      109
Name: Place, dtype: object

The dtype should be int64 but it is showing object. So we have to check the values.

In [411]:
df.Place.unique()

array(['23', '1', '34', '5', '29', '19', '10', '21', '11', '9', '30', '4',
       '17', '22', '31', '7', '18', '14', '32', '16', '28', '15', '6',
       '13', '24', 'Sobanapuram', '26', '20', '35', 'Naganallur', '2'],
      dtype=object)

There are 2 values 'Sobanapuram' and 'Naganallur'. Now we replace them with correct village numbers.

In [412]:
dic={
     'Sobanapuram': '10',
    'Naganallur': '29'
}
df.replace(dic,inplace=True)
df.Place.unique()

array(['23', '1', '34', '5', '29', '19', '10', '21', '11', '9', '30', '4',
       '17', '22', '31', '7', '18', '14', '32', '16', '28', '15', '6',
       '13', '24', '26', '20', '35', '2'], dtype=object)

### Father's and Mother's details

In [413]:
print(df.Faage.describe())
df.Faage.unique()

count    342.000000
mean      44.274854
std        6.676910
min       -1.000000
25%       41.000000
50%       44.000000
75%       48.000000
max       61.000000
Name: Faage, dtype: float64


array([51., 38., 44., 37., 36., 40., 39., 35., 45., 41., -1., 42., 43.,
       32., 55., 47., 48., 46., 50., 54., 34., 58., 61., 53., 49., nan,
       52., 60., 57.])

In [414]:
print(df.Moage.describe())
df.Moage.unique()

count    343.000000
mean      37.705539
std        5.773315
min       -1.000000
25%       34.000000
50%       38.000000
75%       40.000000
max       87.000000
Name: Moage, dtype: float64


array([40, 32, 31, 34, 33, 38, 35, 36, 29, 30, 39, 37, 87, 27, 28, 46, 42,
       44, 43, 41, 45, 48, 50, 47, 49, -1, 56], dtype=int64)

In [415]:
print(df.Faedu.describe())
df.Faedu.unique()

count    343.000000
mean       5.379009
std        1.734353
min       -1.000000
25%        4.500000
50%        5.000000
75%        6.000000
max       10.000000
Name: Faedu, dtype: float64


array([ 5,  6,  7, 10,  4, -1,  3,  9,  1,  8,  2], dtype=int64)

In [416]:
print(df.Moedu.describe())
df.Moedu.unique()

count    343.000000
mean       5.542274
std        1.462131
min        1.000000
25%        5.000000
50%        6.000000
75%        6.000000
max       10.000000
Name: Moedu, dtype: float64


array([ 5,  6,  7,  8,  4,  9,  1,  2,  3, 10], dtype=int64)

In [417]:
print(df.Faocc.describe())
df.Faocc.unique()

count    343.000000
mean       3.443149
std        1.656673
min       -1.000000
25%        2.000000
50%        3.000000
75%        5.000000
max        7.000000
Name: Faocc, dtype: float64


array([ 3,  2,  6,  1,  5, -1,  4,  7], dtype=int64)

In [418]:
print(df.Moocc.describe())
df.Moocc.unique()

count     343
unique      8
top         5
freq      208
Name: Moocc, dtype: object


array(['5', '1', '2', '4', '3', '7', '6', 'Employer'], dtype=object)

In [419]:
df.Fasal.describe()

count    343.000000
mean       4.075802
std        2.548953
min        0.000000
25%        2.000000
50%        4.000000
75%        5.000000
max       13.000000
Name: Fasal, dtype: float64

In [420]:
df.Mosal.describe()

count    343.000000
mean       1.291545
std        1.917026
min        0.000000
25%        0.000000
50%        1.000000
75%        2.000000
max       13.000000
Name: Mosal, dtype: float64

There's one NaN value in Father's Age column, Others columns are clean.

In [421]:
print(df.Stay.describe())
df.Stay.unique()

count    343.000000
mean       1.058309
std        0.269469
min        1.000000
25%        1.000000
50%        1.000000
75%        1.000000
max        3.000000
Name: Stay, dtype: float64


array([1, 2, 3], dtype=int64)

In [422]:
print(df.Famtyp.describe())
df.Famtyp.unique()

count     340
unique     14
top         1
freq      204
Name: Famtyp, dtype: object


array(['1', '3', '2', '4', 'Two parents with two child', '5', 'Parents ',
       '1 father 1 mother ', nan, "Mother's and father two child",
       "Parent's (mother and father)and with siblings (elder sister and younger brother)",
       'Parent, one elder sister, and me',
       'Live with father and mother and one child',
       'Joint family consists of 11 members',
       'They have father mother an two children'], dtype=object)

Only 5 values are to be assigned in this column i.e. 1 to 5, But we have lots of unnecessary values which have to grouped with the 5 values.


In [423]:
dic={
    'Two parents with two child': '1',
    'Parents ':'1',
    '1 father 1 mother ':  '1',
    "Mother's and father two child" : '1',
    "Parent's (mother and father)and with siblings (elder sister and younger brother)" : '4', 
    'Parent, one elder sister, and me' : '1', 
    'Live with father and mother and one child' : '1',
    'Joint family consists of 11 members': '4',
    'They have father mother an two children': '1'
}

In [424]:
df.replace(dic,inplace=True)
print(df.Famtyp.describe())
df.Famtyp.unique()

count     340
unique      5
top         1
freq      211
Name: Famtyp, dtype: object


array(['1', '3', '2', '4', '5', nan], dtype=object)

In [425]:
df[df.Famtyp.isnull()][['Famtyp','Count','Sib']]

Famtyp  Count  Sib
118    NaN      4    1
195    NaN     13    3
324    NaN      7    2

There are 3 Nan values. Considering the Count and Sib Columns we can somewhat guess and fill the null values.

In [426]:
df.loc[118,'Famtyp']='1'
df.loc[195,'Famtyp']='4'
df.loc[324,'Famtyp']='4'

In [427]:
print(df.Famtyp.describe())
df.Famtyp.unique()

count     343
unique      5
top         1
freq      212
Name: Famtyp, dtype: object


array(['1', '3', '2', '4', '5'], dtype=object)

NaN values have been replaced successfully.

In [428]:
print(df.Count.describe())
df.Count.unique()

count    343.000000
mean       4.451895
std        1.183409
min        2.000000
25%        4.000000
50%        4.000000
75%        5.000000
max       13.000000
Name: Count, dtype: float64


array([ 3,  4,  6,  5,  7,  2,  8, 13, 11], dtype=int64)

In [429]:
print(df.Pamas.describe())
df.Pamas.unique()

count    343.000000
mean       1.883382
std        0.560219
min        1.000000
25%        2.000000
50%        2.000000
75%        2.000000
max        8.000000
Name: Pamas, dtype: float64


array([2, 1, 8, 7], dtype=int64)

In [430]:
df[(df.Pamas==8 )| (df.Pamas ==7)][['Count','Pamas']]

Count  Pamas
146      8      8
324      7      7

These 2 students must have accidentally entered the same values for the columns Count and Pamas, we can't guess the correct values as there is no sufficient information about there parent's relationship. Hence we give them someother value which is not in Pamas. We'll use 0.

In [431]:
df['Pamas'].replace({8: 0,7: 0},inplace=True)

In [432]:
print(df.Pamas.describe())
df.Pamas.unique()

count    343.00000
mean       1.83965
std        0.38305
min        0.00000
25%        2.00000
50%        2.00000
75%        2.00000
max        2.00000
Name: Pamas, dtype: float64


array([2, 1, 0], dtype=int64)

In [433]:
print(df.Famrel.describe())
df.Famrel.unique()

count    343.000000
mean       4.454810
std        1.117445
min        1.000000
25%        5.000000
50%        5.000000
75%        5.000000
max        5.000000
Name: Famrel, dtype: float64


array([5, 1, 2, 3, 4], dtype=int64)

In [434]:
print(df.Loved.describe())
df.Loved.unique()

count    343.000000
mean       1.061224
std        0.313974
min        1.000000
25%        1.000000
50%        1.000000
75%        1.000000
max        3.000000
Name: Loved, dtype: float64


array([1, 3, 2], dtype=int64)

In [435]:
print(df.Mosup.describe())
df.Mosup.unique()

count    343.000000
mean       4.338192
std        1.233985
min        1.000000
25%        4.000000
50%        5.000000
75%        5.000000
max        5.000000
Name: Mosup, dtype: float64


array([5, 1, 4, 2, 3], dtype=int64)

In [436]:
print(df.Fasup.describe())
df.Fasup.unique()

count    343.000000
mean       4.262391
std        1.331561
min        1.000000
25%        4.000000
50%        5.000000
75%        5.000000
max        5.000000
Name: Fasup, dtype: float64


array([5, 1, 4, 2, 3], dtype=int64)

In [437]:
print(df.Sib.describe())
df.Sib.unique()

count    343.000000
mean       1.183673
std        0.652096
min        0.000000
25%        1.000000
50%        1.000000
75%        1.000000
max        3.000000
Name: Sib, dtype: float64


array([0, 1, 3, 2], dtype=int64)

In [438]:
print(df..describe())
df..unique()

AttributeError: 'DataFrame' object has no attribute 'Sibtyp'

In [381]:
print(df.Work.describe())
df.Work.unique()

count     343
unique      6
top        No
freq      302
Name: Work, dtype: object


array(['No', 'Yes', 'Yes ', 'no', 'No ', 'N0'], dtype=object)

In [382]:
dic={
    "Yes":1,
    "Yes ":1,
    'No':2, 
    'no':2, 
    'No ':2,
    'N0':2
}

In [384]:
df.Work.replace(dic,inplace=True)
df.Work.unique()

array([2, 1], dtype=int64)

In [321]:
print(df..describe())
df..unique()

count    343.000000
mean       1.058309
std        0.269469
min        1.000000
25%        1.000000
50%        1.000000
75%        1.000000
max        3.000000
Name: Stay, dtype: float64


array([1, 2, 3], dtype=int64)

In [321]:
print(df..describe())
df..unique()

count    343.000000
mean       1.058309
std        0.269469
min        1.000000
25%        1.000000
50%        1.000000
75%        1.000000
max        3.000000
Name: Stay, dtype: float64


array([1, 2, 3], dtype=int64)

In [321]:
print(df..describe())
df..unique()

count    343.000000
mean       1.058309
std        0.269469
min        1.000000
25%        1.000000
50%        1.000000
75%        1.000000
max        3.000000
Name: Stay, dtype: float64


array([1, 2, 3], dtype=int64)